tests/frontend/test_tdmprogram.py

```diff
@@ -0,0 +1,425 @@
+    alpha = [0] * 4
+    phi = [0] * 4
+    theta = [0] * 4
+    np.random.seed(42)
+    x = singleloop(sq_r, alpha, phi, theta, copies)
+    np.random.seed(42)
+    y = singleloop(sq_r, alpha, phi, theta, copies, shift=1)
+    assert np.allclose(x, y)
+
+
+def test_str_tdm_method():
+    """Testing the string method"""
+    prog = tdmprogram.TDMProgram(N=1)
+    assert prog.__str__() == "<TDMProgram: concurrent modes=1, time bins=0, spatial modes=0>"
+
+
+def test_epr():
+    """Generates an EPR state and checks that the correct correlations (noise reductions) are observed
+    from the samples"""
+    np.random.seed(42)
+    sq_r = 1.0
+    c = 4
+    copies = 200
+
+    # This will generate c EPRstates per copy. I chose c = 4 because it allows us to make 4 EPR pairs per copy that can each be measured in different basis permutations.
+    alpha = [np.pi / 4, 0] * c
+    phi = [0, np.pi / 2] * c
+
+    # Measurement of 4 subsequent EPR states in XX, XP, PX, PP to investigate nearest-neighbour correlations in all basis permutations
+    theta = [0, 0] + [0, np.pi / 2] + [np.pi / 2, 0] + [np.pi / 2, np.pi / 2]  #
+
+    x = singleloop(sq_r, alpha, phi, theta, copies)
+
+    X0 = x[0::8]
+    X1 = x[1::8]
+    X2 = x[2::8]
+    P0 = x[3::8]
+    P1 = x[4::8]
+    X3 = x[5::8]
+    P2 = x[6::8]
+    P3 = x[7::8]
+    atol = 5 / np.sqrt(copies)
+    minusstdX1X0 = (X1 - X0).std() / np.sqrt(2)
+    plusstdX1X0 = (X1 + X0).std() / np.sqrt(2)
+    squeezed_std = np.exp(-sq_r)
+    assert np.allclose(minusstdX1X0, squeezed_std, atol=atol)
+    assert np.allclose(plusstdX1X0, 1 / squeezed_std, atol=atol)
+    minusstdP2P3 = (P2 - P3).std() / np.sqrt(2)
+    plusstdP2P3 = (P2 + P3).std() / np.sqrt(2)
+    assert np.allclose(minusstdP2P3, 1 / squeezed_std, atol=atol)
+    assert np.allclose(plusstdP2P3, squeezed_std, atol=atol)
+    minusstdP0X2 = (P0 - X2).std()
+    plusstdP0X2 = (P0 + X2).std()
+    expected = 2 * np.sinh(sq_r) ** 2
+    assert np.allclose(minusstdP0X2, expected, atol=atol)
+    assert np.allclose(plusstdP0X2, expected, atol=atol)
+    minusstdX3P1 = (X3 - P1).std()
+    plusstdX3P1 = (X3 + P1).std()
+    assert np.allclose(minusstdX3P1, expected, atol=atol)
+    assert np.allclose(plusstdX3P1, expected, atol=atol)
+
+
+def test_ghz():
+    """Generates a GHZ state and checks that the correct correlations (noise reductions) are observed
+    from the samples
+    See Eq. 5 of https://advances.sciencemag.org/content/5/5/eaaw4530
+    """
+    # Set up the circuit
+    np.random.seed(42)
+    n = 10
+    vac_modes = 1
+    copies = 1000
+    sq_r = 5
+    alpha = [np.arccos(np.sqrt(1 / (n - i + 1))) if i != n + 1 else 0 for i in range(n + vac_modes)]
+    alpha[0] = 0.0
+    phi = [0] * (n + vac_modes)
+    phi[0] = np.pi / 2
+
+    # Measuring X nullifier
+    theta = [0] * (n + vac_modes)
+    samples_X = singleloop(sq_r, alpha, phi, theta, copies)
+    reshaped_samples_X = np.array(samples_X).reshape([copies, n + vac_modes])
```

Comment:
What does `reshape([copies, n + vac_modes])` do here?

Reply:
it turns the vector of length `copies * (n+vac_modes)` into a matrix of shape `[copies, n+vac_modes]`

Reply:
Love it! Does this work for >1 spatial modes too?